apps/reviews-api/data/get_review.go

```diff
@@ -2,48 +2,149 @@ package data
+		}
+		log.Println("exec find reviews query: ", err)
+		return nil, fiber.ErrInternalServerError
 	}
 
-	reviews, err := client.Review.FindMany(
-		db.Review.CourseID.Equals(course.ID),
+	return reviews, nil
+}
+
+func getMyReview(ctx context.Context, client *db.PrismaClient, courseID int, userID string, myReviewIDChan chan<- *db.ReviewModel) {
+	myReview, err := client.Review.FindFirst(
+		db.Review.CourseID.Equals(courseID),
+		db.Review.UserID.Equals(userID),
 	).With(
 		db.Review.Course.Fetch(),
-        db.Review.Profile.Fetch(),
+		db.Review.Profile.Fetch(),
 	).Exec(ctx)
 	if err != nil {
-		return nil, errors.New(fmt.Sprintf("GetAllReviews: find reviews: %v", err))
+		if !errors.Is(err, db.ErrNotFound) {
+			log.Println("exec find my reviews query: ", err)
+		}
+		myReviewIDChan <- nil
+		return
 	}
 
-	var reviewJSONResponses []ReviewJSONResponse
+	myReviewIDChan <- myReview
+}
+
+func parseReviewJSONResponse(reviews []db.ReviewModel, myReview *db.ReviewModel) []ReviewJSONResponse {
+	myReviewID := 0 // prevent nil pointer dereference in case myReview does not exist
+	if myReview != nil {
+		reviews = append([]db.ReviewModel{*myReview}, reviews...) // add myReview to the beginning of reviews
+		myReviewID = myReview.ID
+	}
+
+	reviewJSONResponses := []ReviewJSONResponse{}
 	for _, review := range reviews {
+		profile, ok := review.Profile()
+		if !ok {
+			log.Printf("failed to get profile for review id %d", review.ID)
+			continue
+		}
+
 		reviewJSONResponses = append(reviewJSONResponses, ReviewJSONResponse{
 			ID:           review.ID,
 			AcademicYear: review.AcademicYear,
 			Description:  review.Description,
+			IsOwner:      review.ID == myReviewID,
 			Rating:       review.Rating,
+			Status:       review.Status,
 			Votes:        review.Votes,
 			Course: CourseJSONResponse{
 				ID:   review.Course().ID,
 				Code: review.Course().Code,
 			},
+			Profile: UserJSONResponse{
+				ID: profile.ID,
+			},
 		})
 	}
-	return reviewJSONResponses, nil
+
+	return reviewJSONResponses
+}
+
+func GetCourseReviews(ctx context.Context, client *db.PrismaClient, courseCode, userID string, pageSize, pageNumber int) (*utils.Pagination, error) {
+	myReviewChan := make(chan *db.ReviewModel)
+	defer close(myReviewChan)
+
+	courseID, err := getCourseID(ctx, client, courseCode)
+	if err != nil {
+		return nil, err
+	}
+
+	// run getMyReview concurrently with getReviews
+	go getMyReview(ctx, client, courseID, userID, myReviewChan)
+	rawReviews, err := getReviews(ctx, client, userID, courseID)
+	if err != nil {
+		return nil, err
+	}
+
+	totalNumberOfItems := len(rawReviews)
+
+	// force only 2 reviews to be retrived for non-active users
+	if !isActiveUser(ctx, client, userID) {
+		pageSize = 2
+		pageNumber = 1
+		totalNumberOfItems = maxReviewsForNonActiveUsers
+	}
+
+	myReview := <-myReviewChan
+	if myReview != nil {
+		totalNumberOfItems++
+	}
+
+	start, end := utils.CalculateStartEnd(rawReviews, pageSize, pageNumber)
```

Comment:
Couldn't we just use LIMIT and OFFSET in the query?

Reply:
I meant to press comment but GitHub :shrug:

Reply:
> Couldn't we just use LIMIT and OFFSET in the query?

I retrieved the whole thing to determine the total number of pages that can be sent along the response.

This way the frontend can do stuff like showing the total number of pages at the bottom, or know whether or not they can fetch more data

Using limit and offset might not allow this idea (or is there any other way to do it?)

Reply:
Prolly better to just have a function like QueryTotalCountOfReviews and call that and return it along with the result, something like `SELECT COUNT(*)` using COUNT is much faster than querying every row, and the original query can be made into LIMIT + OFFSET

Reply:
okie dokie

prisma go apparently does not support COUNT, so imma go with raw query XD

Reply:
> okie dokie
> 
> prisma go apparently does not support COUNT, so imma go with raw query XD

Wowwww no COUNT support?! Go kinda laaaame~? 😃

Reply:
THE LIB IS ACTUALLY ASS!!!